make aoti_torch_empty_strided support creating incontiguous tensor

Gasoonjia opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #15282
* #15241
* __->__ #15228

This diff modifies the `aoti_torch_empty_strided` function to support the creation of incontiguous tensors.

To achieve it, this diff:
1. update the way to calculate the memory size by using both tensor size and the stride
2. skip stride check in ETensor by adding and checking cmake macro `USE_CUDA_BACKEND` when building with CUDA backend support.

we will soon bring the ETensor check back for every backend after migrating to use slimtensor.

Differential Revision: [D84938258](https://our.internmc.facebook.com/intern/diff/D84938258/)

Gasoonjia committed:
make aoti_torch_empty_strided support creating incontiguous tensor

This diff modifies the `aoti_torch_empty_strided` function to support the creation of incontiguous tensors.

To achieve it, this diff:
1. update the way to calculate the memory size by using both tensor size and the stride
2. skip stride check in ETensor by adding and checking cmake macro `USE_CUDA_BACKEND` when building with CUDA backend support.

we will soon bring the ETensor check back for every backend after migrating to use slimtensor.

Differential Revision: [D84938258](https://our.internmc.facebook.com/intern/diff/D84938258/)

[ghstack-poisoned]

Gasoonjia committed:
Update on "make aoti_torch_empty_strided support creating incontiguous tensor"


This diff modifies the `aoti_torch_empty_strided` function to support the creation of incontiguous tensors.

To achieve it, this diff:
1. update the way to calculate the memory size by using both tensor size and the stride
2. skip stride check in ETensor by adding and checking cmake macro `USE_CUDA_BACKEND` when building with CUDA backend support.

we will soon bring the ETensor check back for every backend after migrating to use slimtensor.

Differential Revision: [D84938258](https://our.internmc.facebook.com/intern/diff/D84938258/)

[ghstack-poisoned]

Gasoonjia committed:
Update on "make aoti_torch_empty_strided support creating incontiguous tensor"


This diff modifies the `aoti_torch_empty_strided` function to support the creation of incontiguous tensors.

To achieve it, this diff:
1. update the way to calculate the memory size by using both tensor size and the stride
2. skip stride check in ETensor by adding and checking cmake macro `USE_CUDA_BACKEND` when building with CUDA backend support.

we will soon bring the ETensor check back for every backend after migrating to use slimtensor.

Differential Revision: [D84938258](https://our.internmc.facebook.com/intern/diff/D84938258/)

[ghstack-poisoned]

JacobSzwejbka commented on extension/tensor/tensor_ptr.cpp:
```diff
@@ -79,6 +79,11 @@ TensorPtr make_tensor_ptr(
+// Skip stride calculation and incontiguous tensor check for CUDA backend since
+// AOTI-CUDA handles both contiguous and incontiguous tensors. This will be
+// removed after SlimTensor migration.
+#ifndef USE_CUDA_BACKEND
```
😭

Gasoonjia committed:
Update on "make aoti_torch_empty_strided support creating incontiguous tensor"


This diff modifies the `aoti_torch_empty_strided` function to support the creation of incontiguous tensors.

To achieve it, this diff:
1. update the way to calculate the memory size by using both tensor size and the stride
2. skip stride check in ETensor by adding and checking cmake macro `USE_CUDA_BACKEND` when building with CUDA backend support.

we will soon bring the ETensor check back for every backend after migrating to use slimtensor.

Differential Revision: [D84938258](https://our.internmc.facebook.com/intern/diff/D84938258/)

[ghstack-poisoned]

Gasoonjia committed:
Update on "make aoti_torch_empty_strided support creating incontiguous tensor"


This diff modifies the `aoti_torch_empty_strided` function to support the creation of incontiguous tensors.

To achieve it, this diff:
1. update the way to calculate the memory size by using both tensor size and the stride
2. skip stride check in ETensor by adding and checking cmake macro `USE_CUDA_BACKEND` when building with CUDA backend support.

we will soon bring the ETensor check back for every backend after migrating to use slimtensor.

Differential Revision: [D84938258](https://our.internmc.facebook.com/intern/diff/D84938258/)

[ghstack-poisoned]

larryliu0820 commented on extension/tensor/tensor_ptr.cpp:
```diff
@@ -79,6 +79,11 @@ TensorPtr make_tensor_ptr(
+// Skip stride calculation and incontiguous tensor check for CUDA backend since
+// AOTI-CUDA handles both contiguous and incontiguous tensors. This will be
+// removed after SlimTensor migration.
+#ifndef USE_CUDA_BACKEND
```
This is not needed anymore?

Gasoonjia replied:
you are absolutely right. Removed.

Gasoonjia committed:
Update on "make aoti_torch_empty_strided support creating incontiguous tensor"


This diff modifies the `aoti_torch_empty_strided` function to support the creation of incontiguous tensors.

To achieve it, this diff:
1. update the way to calculate the memory size by using both tensor size and the stride
2. skip stride check in ETensor by adding and checking cmake macro `USE_CUDA_BACKEND` when building with CUDA backend support.

we will soon bring the ETensor check back for every backend after migrating to use slimtensor.

Differential Revision: [D84938258](https://our.internmc.facebook.com/intern/diff/D84938258/)

[ghstack-poisoned]

Gasoonjia committed:
Update on "make aoti_torch_empty_strided support creating incontiguous tensor"


This diff modifies the `aoti_torch_empty_strided` function to support the creation of incontiguous tensors.

To achieve it, this diff:
1. update the way to calculate the memory size by using both tensor size and the stride
2. skip stride check in ETensor by adding and checking cmake macro `USE_CUDA_BACKEND` when building with CUDA backend support.

we will soon bring the ETensor check back for every backend after migrating to use slimtensor.

Differential Revision: [D84938258](https://our.internmc.facebook.com/intern/diff/D84938258/)

[ghstack-poisoned]

Gasoonjia committed:
Update on "make aoti_torch_empty_strided support creating incontiguous tensor"


This diff modifies the `aoti_torch_empty_strided` function to support the creation of incontiguous tensors.

To achieve it, this diff:
1. update the way to calculate the memory size by using both tensor size and the stride
2. skip stride check in ETensor by adding and checking cmake macro `USE_CUDA_BACKEND` when building with CUDA backend support.

we will soon bring the ETensor check back for every backend after migrating to use slimtensor.

Differential Revision: [D84938258](https://our.internmc.facebook.com/intern/diff/D84938258/)

[ghstack-poisoned]

Gasoonjia committed:
Update on "make aoti_torch_empty_strided support creating incontiguous tensor"


This diff modifies the `aoti_torch_empty_strided` function to support the creation of incontiguous tensors.

To achieve it, this diff:
1. update the way to calculate the memory size by using both tensor size and the stride
2. skip stride check in ETensor by adding and checking cmake macro `USE_CUDA_BACKEND` when building with CUDA backend support.

we will soon bring the ETensor check back for every backend after migrating to use slimtensor.

Differential Revision: [D84938258](https://our.internmc.facebook.com/intern/diff/D84938258/)

[ghstack-poisoned]
